cognite/client/data_classes/data_modeling/typed_instances.py

```diff
@@ -317,27 +277,32 @@ def _get_properties_by_name(cls: type) -> dict[str, PropertyOptions]:
 )
 
 
+@lru_cache(32)
+def _get_decriptor_property_name_mapping(typed_cls: type[_TypedInstance]) -> MappingProxyType[str, str]:
+    return MappingProxyType(
+        {v.name: k for cls in typed_cls.__mro__[:-1] for k, v in cls.__dict__.items() if isinstance(v, PropertyOptions)}
+    )
+
+
 def _dump_properties(
     obj: object, camel_case: bool, instance_properties: frozenset[str], use_attribute_name: bool = False
 ) -> dict[str, Any]:
     properties: dict[str, str | int | float | bool | dict | list] = {}
-    properties_by_name = _get_properties_by_name(type(obj))
-    if use_attribute_name:
-        attribute_by_property = {v.name: k for k, v in properties_by_name.items()}
-    else:
-        attribute_by_property = {}
+    attr_name_map = _get_decriptor_property_name_mapping(type(obj)) if use_attribute_name else {}  # type: ignore [arg-type]
+
     for key, value in vars(obj).items():
         if key in instance_properties or value is None:
             continue
         if key.startswith("__"):
             key = key[2:]
 
-        if key in properties_by_name:
-            key = cast(str, properties_by_name[key].name)
-        if use_attribute_name and key in attribute_by_property:
-            key = attribute_by_property[key]
+        if use_attribute_name and key in attr_name_map:
+            # We must rstrip any underscores on the attribute name because they may have been added due
+            # to a possible namespace conflict for a property identifier. (If not, DMS will block it as
+            # an invalid identifier anyway)
+            key = attr_name_map[key].rstrip("_")
```

Comment:
I'll add a test for this using CogniteAsset with both `type` and `type_`, but have to run to kindergarten now 😅